bridges/snowbridge/pallets/outbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+
+	/// The current nonce for the messages
+	#[pallet::storage]
+	pub type Nonce<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+	/// Pending orders to relay
+	#[pallet::storage]
+	pub type PendingOrders<T: Config> =
+		StorageMap<_, Identity, u64, PendingOrder<BlockNumberFor<T>>, OptionQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+	where
+		T::AccountId: AsRef<[u8]>,
+	{
+		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+			// Remove storage from previous block
+			Messages::<T>::kill();
+			MessageLeaves::<T>::kill();
+			// Reserve some weight for the `on_finalize` handler
+			T::WeightInfo::commit()
+		}
+
+		fn on_finalize(_: BlockNumberFor<T>) {
+			Self::commit();
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T::AccountId: From<[u8; 32]>,
+	{
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::submit_delivery_receipt())]
+		pub fn submit_delivery_receipt(
+			origin: OriginFor<T>,
+			event: Box<EventProof>,
+		) -> DispatchResult {
+			let relayer = ensure_signed(origin)?;
+
+			// submit message to verifier for verification
+			T::Verifier::verify(&event.event_log, &event.proof)
+				.map_err(|e| Error::<T>::Verification(e))?;
+
+			let receipt = DeliveryReceiptOf::<T>::try_from(&event.event_log)
+				.map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			Self::process_delivery_receipt(relayer, receipt)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Generate a messages commitment and insert it into the header digest
+		pub(crate) fn commit() {
+			let count = MessageLeaves::<T>::decode_len().unwrap_or_default() as u64;
+			if count == 0 {
+				return;
+			}
+
+			// Create merkle root of messages
+			let root = merkle_root::<<T as Config>::Hashing, _>(MessageLeaves::<T>::stream_iter());
+
+			let digest_item: DigestItem = CustomDigestItem::SnowbridgeV2(root).into();
+
+			// Insert merkle root into the header digest
+			<frame_system::Pallet<T>>::deposit_log(digest_item);
+
+			Self::deposit_event(Event::MessagesCommitted { root, count });
+		}
+
+		/// Process a message delivered by the MessageQueue pallet
+		pub(crate) fn do_process_message(
+			_: ProcessMessageOriginOf<T>,
+			mut message: &[u8],
+		) -> Result<bool, ProcessMessageError> {
+			use ProcessMessageError::*;
+
+			// Yield if the maximum number of messages has been processed this block.
+			// This ensures that the weight of `on_finalize` has a known maximum bound.
+			ensure!(
+				MessageLeaves::<T>::decode_len().unwrap_or(0) <
+					T::MaxMessagesPerBlock::get() as usize,
+				Yield
+			);
+
+			let nonce = Nonce::<T>::get();
+
+			// Decode bytes into Message and
+			// a. Convert to OutboundMessage and save into Messages
+			// b. Convert to committed hash and save into MessageLeaves
+			// c. Save nonce&fee into PendingOrders
+			let message: Message = Message::decode(&mut message).map_err(|_| Corrupt)?;
+			let commands: Vec<OutboundCommandWrapper> = message
+				.commands
+				.clone()
```

Comment:
If these commands are not needed anymore in the code that follows, I think we can avoid cloning by just doing `core::mem::take(&mut message.commands)`.

Or we can deconstruct the message above:
```
			let Message { origin, id, fee, commands, .. } =
				Message::decode(&mut message).map_err(|_| Corrupt)?;
```

Reply:
https://github.com/paritytech/polkadot-sdk/pull/7402/commits/204962fab10268cd2e84d95593d07e8e6de94fb4